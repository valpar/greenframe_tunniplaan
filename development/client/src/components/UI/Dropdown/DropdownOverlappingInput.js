import { useEffect, useState } from "react";
import TooltipLarge from "../Tooltip/TooltipLarge";

const DropdownOverlappingInput = (props) => {
  const [options, setOptions] = useState();
  const [showTooltip, setShowTooltip] = useState(false);
  let timer;
  useEffect(() => {
    setOptions(props.options);
  }, []);

  const { isTabletOrMobile } = props;

  const [showOptions, setShowOptions] = useState(false);

  const inputChangeHandler = (event) => {
    event.preventDefault();
    setOptions(
      props.options.filter((e) => e.label.includes(event.target.value))
    );
    props.onChange(event.target.value);
  };
  const showOptionsHandler = () => {
    if (props.showOptions)
      setShowOptions((prevState) => (prevState = !prevState));
  };
  const optionClickHandler = (event) => {
    event.preventDefault();
    props.onChange({
      id: props.index,
      value: event.target.innerText,
    });
    setShowOptions((prevState) => (prevState = !prevState));
  };
  const loseFocusHandler = () => {
    timer = setTimeout(() => {
      setShowOptions((prevState) => (prevState = false));
    }, 200);
  };

  useEffect(() => {
    return () => clearTimeout(timer);
  }, []);

  const mouseEnterHandler = () => {
    setShowTooltip(true);
  };
  const mouseLeaveHandler = () => {
    setShowTooltip(false);
  };

  return (
    <div className="relative flex flex-col">
      {(props.hasError || (props.errorMessage > "" && showTooltip)) && (
        <div
          className={`absolute left-1/2 -translate-x-1/2 bottom-14 w-full max-w-xs`}
        >
          <TooltipLarge message={props.errorMessage} />
        </div>
      )}
      <label
        htmlFor="input"
        className={`absolute left-0 ml-2 text-xs -top-2 ${
          props.value
            ? "text-gray-500 bg-white px-1"
            : "text-gray-400 bg-transparent"
        } duration-200`}
      >
        {props.value ? props.placeholder : ""}
      </label>
      <input
        data-testid="option-1" // for testing purposes
        onMouseEnter={mouseEnterHandler}
        onMouseLeave={mouseLeaveHandler}
        onClick={showOptionsHandler}
        placeholder={props.placeholder}
        onChange={inputChangeHandler}
        type={props.type ? props.type : "text"}
        name={props.name ? props.name : ""}
        value={props.value ? props.value : ""}
        readOnly={props.readOnly ? true : false}
        onBlur={loseFocusHandler}
        autoComplete="off"
        className={`p-2 border w-full ${
          props.hasError || props.errorMessage
            ? "border-red-500"
            : "border-borderGray "
        } rounded-none font-normal focus:outline-none focus:border-black`}
      />

      {props.showOptions && showOptions && options.length !== 0 && (
        <div className="absolute top-12 w-full bg-white border border-black/50 shadow z-20">
          {options.map((e, i) => {
            return (
              <div
                key={i}
                value={e.value}
                onClick={optionClickHandler}
                className="p-2 hover:bg-sky-100 duration-150"
              >
                {e.label}
              </div>
            );
          })}
        </div>
      )}
    </div>
  );
};

export default DropdownOverlappingInput;
