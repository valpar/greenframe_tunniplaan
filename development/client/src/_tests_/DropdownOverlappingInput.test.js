import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import DropdownOverlappingInput from '../components/UI/Dropdown/DropdownOverlappingInput.js';


describe('DropdownOverlappingInput', () => {
  it("matches snapshot", () => {
    const tree = renderer.create(<DropdownOverlappingInput />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('filters options based on user input', () => {
    const options = [
      { label: 'Option 1', value: 'Option 1' },
      { label: 'Option 2', value: 'Option 2' },
      { label: 'Option 3', value: 'Option 3' },
    ];

    // Define a mock onChange function
    const mockOnChange = jest.fn();

    render(
      <DropdownOverlappingInput
        placeholder="Test Placeholder"
        options={options}
        onChange={mockOnChange} // Pass the mock onChange function
      />
    );

    const input = screen.getByPlaceholderText('Test Placeholder');

    // Simulate user input
    fireEvent.change(input, { target: { value: 'Option 1' } });

    // Query for the element using the test ID
    const option1 = screen.getByTestId('option-1');

    // Assert that the filtered option is displayed
    expect(option1).toBeInTheDocument();

    // Assert that the onChange function was called
    expect(mockOnChange).toHaveBeenCalledWith('Option 1');
  });
  
  
});
