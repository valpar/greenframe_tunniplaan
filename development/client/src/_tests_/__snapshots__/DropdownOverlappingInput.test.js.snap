// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownOverlappingInput matches snapshot 1`] = `
<div
  className="relative flex flex-col"
>
  <label
    className="absolute left-0 ml-2 text-xs -top-2 text-gray-400 bg-transparent duration-200"
    htmlFor="input"
  />
  <input
    autoComplete="off"
    className="p-2 border w-full border-borderGray  rounded-none font-normal focus:outline-none focus:border-black"
    data-testid="option-1"
    name=""
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;
